adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java

```diff
@@ -1038,6 +1069,86 @@ public AdyenPaymentService getAdyenPaymentService() {
+            customer = getCheckoutCustomerStrategy().getCurrentUserForCheckout();
+        }
+        //This will be used to check status later
+
+        String serviceId = request.getAttribute("originalServiceId").toString();
+        TerminalAPIResponse terminalApiResponse = getAdyenPaymentService().sendSyncPosPaymentRequest(cartData, customer, serviceId);
+        ResultType resultType = TerminalAPIUtil.getPaymentResultFromStatusOrPaymentResponse(terminalApiResponse);
+
+        if (ResultType.SUCCESS == resultType) {
+            PaymentsResponse paymentsResponse = getPosPaymentResponseConverter().convert(terminalApiResponse.getSaleToPOIResponse());
+            String posReceipt = TerminalAPIUtil.getReceiptFromPaymentResponse(terminalApiResponse);
+            if (StringUtils.isNotEmpty(posReceipt)) {
+                paymentsResponse.putAdditionalDataItem("pos.receipt", posReceipt);
+            }
+            return createAuthorizedOrder(paymentsResponse);
+        }
+        throw new AdyenNonAuthorizedPaymentException(terminalApiResponse);
+    }
+
+    /**
+     * Ininiate POS Payment using Adyen Terminal API
+     */
+    @Override
+    public OrderData checkPosPaymentStatus(HttpServletRequest request, CartData cartData) throws Exception {
+
+        String originalServiceId = request.getAttribute("originalServiceId").toString();
+        TerminalAPIResponse terminalApiResponse = getAdyenPaymentService().sendSyncPosStatusRequest(cartData, originalServiceId);
+        ResultType statusResult = TerminalAPIUtil.getStatusResultFromStatusResponse(terminalApiResponse);
+
+        if (statusResult != null) {
+            if (statusResult == ResultType.SUCCESS) {
+                //this will be success even if payment is failed. because this belongs to status call not payment call
+                ResultType paymentResult = TerminalAPIUtil.getPaymentResultFromStatusOrPaymentResponse(terminalApiResponse);
+                if (paymentResult == ResultType.SUCCESS) {
+                    PaymentsResponse paymentsResponse = getPosPaymentResponseConverter().convert(terminalApiResponse.getSaleToPOIResponse());
+                    String posReceipt = TerminalAPIUtil.getReceiptFromStatusResponse(terminalApiResponse);
+                    if (StringUtils.isNotEmpty(posReceipt)) {
+                        paymentsResponse.putAdditionalDataItem("pos.receipt", posReceipt);
+                    }
+                    return createAuthorizedOrder(paymentsResponse);
+                } else {
+                    throw new AdyenNonAuthorizedPaymentException(terminalApiResponse);
+                }
+            } else {
+                ErrorConditionType errorCondition = TerminalAPIUtil.getErrorConditionForStatusFromStatusResponse(terminalApiResponse);
+                //If transaction is still in progress, keep retrying in 5 seconds.
+                if (errorCondition == ErrorConditionType.IN_PROGRESS) {
+                    TimeUnit.SECONDS.sleep(5);
```

Comment:
why are we doing a thread sleep here ? Would avoid doing thread sleep at all times and better to do an new request from the front-end.

Reply:
I don't see any problem in using Thread.sleep because we are not explicitly managing threads  . Thread management is being done by framework itself and we are just holding current thread for 5 seconds. Sending the call back to Front end means someone has to trigger the request manually again to check the status. We can discuss more if you want.

Reply:
Using thread sleep is in general a bad design and as well a heavy load to the server so think polling is the better solution here. You can auto poll it from the front-end so no customer interaction is needed. Hava a chat with alessio he did it for magento i believe.

Reply:
Let's merge it for now and discuss this separately with pro/cons of this.